src/OrchardCore.Modules/OrchardCore.Deployment/Controllers/ImportController.cs

```diff
@@ -102,7 +109,55 @@ public async Task<IActionResult> Import(IFormFile importedPackage)
                 _notifier.Error(H["Please add a file to import."]);
             }
 
-            return RedirectToAction("Index");
+            return RedirectToAction(nameof(Index));
+        }
+
+        public async Task<IActionResult> Json()
+        {
+            if (!await _authorizationService.AuthorizeAsync(User, Permissions.Import))
+            {
+                return Forbid();
+            }
+
+            return View();
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> Json(ImportJsonViewModel model)
+        {
+            if (!await _authorizationService.AuthorizeAsync(User, Permissions.Import))
+            {
+                return Forbid();
+            }
+
+            if (!model.Json.IsJson())
+            {
+                ModelState.AddModelError(nameof(model.Json), S["The recipe is written in an incorrect json format."]);
+            }
+
+            if (ModelState.IsValid)
+            {
+                var tempArchiveFolder = PathExtensions.Combine(Path.GetTempPath(), Path.GetRandomFileName());
+
+                try
+                {
+                    Directory.CreateDirectory(tempArchiveFolder);
+                    System.IO.File.WriteAllText(Path.Combine(tempArchiveFolder, "Recipe.json"), model.Json);
+
+                    await _deploymentManager.ImportDeploymentPackageAsync(new PhysicalFileProvider(tempArchiveFolder));
+
+                    _notifier.Success(H["Recipe imported."]);
```

Comment:
return RedirectToAction(nameof(Json));

Reply:
Just to do a return to here on success.
We could have a catch here, to add a model state error, but equally we don't on the import package controller.
One day we will want better error messages from the recipe executor, but that's another pr.
(When you run this on a prod server, where the environment is not development with errors, you just get a 500 response back because development exception handling is turned of)

Reply:
I didn't use `return RedirectToAction(nameof(Json));` because I want the json to be kept in the textarea.

Reply:
So stays even after success? cool, sounds fine to me if you prefer that